token-handler/src/mint_tokens.rs

```diff
@@ -1,79 +1,205 @@
-                }
-
-                arg_buffer.push_arg(operation_token.token_nonce);
-                arg_buffer.push_arg(token_data.creator);
-
-                let output = self.send_raw().call_local_esdt_built_in_function(
-                    self.blockchain().get_gas_left(),
-                    &ManagedBuffer::from(ESDT_NFT_CREATE_FUNC_NAME),
-                    &arg_buffer,
+                let arg_buffer = self.get_nft_create_args(
+                    &operation_token.token_identifier,
+                    &operation_token.token_nonce,
+                    &operation_token.token_data,
                 );
 
-                if let Some(first_result_bytes) = output.try_get(0) {
-                    nonce = first_result_bytes.parse_as_u64().unwrap_or_default()
-                } else {
-                    nonce = 0
-                }
+                nonce = self.call_nft_create_built_in_function(&arg_buffer);
             }
 
-            // self.minted_tokens().push(&operation_token);
-
             output_payments.push(OperationEsdtPayment {
                 token_identifier: operation_token.token_identifier,
                 token_nonce: nonce,
                 token_data: operation_token.token_data,
             });
         }
+
+        self.distribute_payments(&tokens_vec, &opt_transfer_data, &to);
+    }
+
+    fn call_nft_create_built_in_function(&self, arg_buffer: &ManagedArgBuffer<Self::Api>) -> u64 {
+        let output = self.send_raw().call_local_esdt_built_in_function(
+            self.blockchain().get_gas_left(),
+            &ManagedBuffer::from(ESDT_NFT_CREATE_FUNC_NAME),
+            arg_buffer,
+        );
+
+        if let Some(first_result_bytes) = output.try_get(0) {
+            first_result_bytes.parse_as_u64().unwrap_or_default()
+        } else {
+            0
+        }
+    }
+
+    fn get_nft_create_args(
+        &self,
+        token_identifier: &TokenIdentifier<Self::Api>,
+        token_nonce: &u64,
+        token_data: &StolenFromFrameworkEsdtTokenData<Self::Api>,
+    ) -> ManagedArgBuffer<Self::Api> {
+        let mut arg_buffer = ManagedArgBuffer::new();
+        let cloned_token_data = token_data.clone();
+
+        arg_buffer.push_arg(token_identifier);
+        arg_buffer.push_arg(cloned_token_data.amount);
+        arg_buffer.push_arg(cloned_token_data.name);
+        arg_buffer.push_arg(cloned_token_data.royalties);
+        arg_buffer.push_arg(cloned_token_data.hash);
+        arg_buffer.push_arg(cloned_token_data.attributes);
+
+        let uris = token_data.uris.clone();
+
+        if uris.is_empty() {
+            // at least one URI is required, so we push an empty one
+            arg_buffer.push_arg(codec::Empty);
+        } else {
+            // The API function has the last argument as variadic,
+            // so we top-encode each and send as separate argument
+            for uri in &uris {
+                arg_buffer.push_arg(uri);
+            }
+        }
+
```

Comment:
We need one more argument here (before nonce): arg_buffer.push_arg(cloned_token_data.token_type);